File: PYTHON/edit_distance.py
import sys


def edit_distance(str1, str2):
    if len(str1) > len(str2):
        str1, str2 = str2, str1

    distances = range(len(str1) + 1)

    for count1, letter1 in enumerate(str2):
        distances_ = [count1+1]
        
        for count2, letter2 in enumerate(str1):
            if letter2 == letter1:
                distances_.append(distances[count2])
            else:
                distances_.append(1 + min((distances[count2], distances[count2 + 1], distances_[-1])))
    
        distances = distances_

    cost = distances[-1]
    return cost


if __name__ == "__main__":
    if len(sys.argv) == 3:
        str1 = sys.argv[1]
        str2 = sys.argv[2]
    else:
        str1 = "Cat"
        str2 = "Rats"
    
    no_operations = edit_distance(str1, str2)
    print(f"Levenshtein distance between < {str1} > and < {str2} >")
    print(f"Number of operations required : {no_operations}")

File: CPP/edit-distance.c
int min(int i,int j,int k){
    if(i<j && i<k){
        return i;
    }
    else if(j<i && j<k){
        return j;
    }
    else{
        return k;
    }
}
int main() {
    int num;
    scanf("%d",&num);
    while(num>0){
        num-=1;
        int i,j,k,l,m,n,l1,l2;
        char str1[1000],str2[1000];
        scanf("%d %d\n",&l1,&l2);
        scanf("%s %s",str1,str2);
        int dp[l1+1][l2+1];
        //memset(dp,0,sizeof(dp));
        for(i=0;i<=l1;i++){
            for(j=0;j<=l2;j++){
                if(i==0){
                    dp[i][j]=j;
                }
                else if(j==0){
                    dp[i][j]=i;
                }
                else if(str1[i-1]==str2[j-1]){
                    dp[i][j]=dp[i-1][j-1];
                }
                else{
                    k=min(dp[i-1][j],dp[i][j-1],dp[i-1][j-1]);
                    dp[i][j]=k+1;
                }
            }
        }
        printf("%d\n",dp[l1][l2]);
    }
}
